Sources/AppcuesKit/Presentation/Traits/TraitComposer.swift

```diff
@@ -120,6 +92,46 @@ internal class TraitComposer: TraitComposing {
 
 @available(iOS 13.0, *)
 extension TraitComposer {
+    class DecomposedTraits {
+        private(set) var allTraitInstances: [ExperienceTrait]
+
+        private(set) var stepDecorators: [StepDecoratingTrait]
+        private(set) var containerCreating: ContainerCreatingTrait?
+        private(set) var containerDecorating: [ContainerDecoratingTrait]
+        private(set) var backdropDecorating: [BackdropDecoratingTrait]
+        private(set) var wrapperCreating: WrapperCreatingTrait?
+        private(set) var presenting: PresentingTrait?
+
+        init(traits: [ExperienceTrait]) {
+            allTraitInstances = traits
+
+            stepDecorators = traits.compactMap { ($0 as? StepDecoratingTrait) }
+            containerCreating = traits.compactMapFirst { ($0 as? ContainerCreatingTrait) }
+            containerDecorating = traits.compactMap { ($0 as? ContainerDecoratingTrait) }
+            backdropDecorating = traits.compactMap { ($0 as? BackdropDecoratingTrait) }
+            wrapperCreating = traits.compactMapFirst { ($0 as? WrapperCreatingTrait) }
+            presenting = traits.compactMapFirst { ($0 as? PresentingTrait) }
+        }
+
+        /// Combine two `DecomposedTrait` instances. For trait types where a single instance is allowed, prefer the current one.
```

Comment:
Yes. This is consistent with how it was:

> The order of precedence [for single trait capabilities] is experience-level traits and then step-level traits, in their order in the experience object.
-  https://github.com/appcues/appcues-ios-sdk/blame/main/Sources/AppcuesKit/AppcuesKit.docc/Traits.md#L120

IIRC my rationale for this decision (I don't think I considered it too closely) in the past was mostly for the case where step-child level traits get hoisted up. It feels a bit wrong that (for example) a PresentingTrait on step 2 of a group overrides the step-group level PresentingTrait.

I can see how it's unintuitive, and having written the above out, it makes less sense than I'd probably originally thought, so I'd be open to changing it. It wouldn't cause any issues given what the builder does today. Do you know off the top of your head how Android does it?

Reply:
Android (cc @andretortolano ) looks like it's currently the other way around, where the `mergeTraits` extension prefers the existing traits (i.e. step level) over the incoming (i.e. group level) https://github.com/appcues/appcues-android-sdk/blob/main/appcues/src/main/java/com/appcues/data/mapper/MapperExt.kt#L10

it gets called during the mapping, like here when a Step is mapped and the higher level traits are passed through https://github.com/appcues/appcues-android-sdk/blob/main/appcues/src/main/java/com/appcues/data/mapper/step/StepMapper.kt#L27

Reply:
The use case I was thinking was the "override" behavior - i.e. the group level defines a green backdrop, but one step inside that group can override and define a red backdrop. Or, the group is skippable, but one step can override and remove the skip. However, those things could probably just as easily be defined at each step and not need such overrides, and maybe that is what you have in mind. I just thought I recalled some discussion of an override capability like this where the lower level trait would take precedence.

Reply:
The decorating traits (ie backdrop) do work that way so overrides are possible. That's captured in the `.append` part, so the order in the array is essentially `[Experience-level, Group-level, Step-level]`, meaning the step-level ones get applied last. The actual implementation of the trait decides what that means, The trait can either just set a single value (say `backdropView.backgroundColor = color`) so that the last one in the list wins, or it could stack them (eg `backdropView.addSubview(ColoredBackdropView(color: color))`). Depending on the trait, either might make sense, but the "last one wins" will be more common.

The "prefer the current one" part is for traits that only can have a single instance (container creating, wrapper creating, presenting). If Android does it the other way around, then I'll update this PR to match that.

And for "Or, the group is skippable, but one step can override and remove the skip." this is tricky since we don't have a way for a step to say it's "not skippable" so in this case, we'd need to only add the skippable trait to the specific steps, rather than having it at the group level and then removing on a specific child.

Reply:
yes, thank you for the explanation on single-instance vs not (decorating) - right. brain waking up slowly after Thanksgiving break. As for the others, I'm not sure if Android is more correct, let's discuss?

Reply:
Ok, I've pushed a few new commits to this branch that
1. Update to prefer the most specific trait capabilities over the more general ones, and
2. Ignore ContainerCreating, WrapperCreating, and Presenting traits on child steps

The changes to allow only a single instance of a given trait type and to propagate experience level and group level traits to the step level will be done separately.

Made a quick documentation update as well so that it's at least correct, but I'm sure there will be a more substantial trait doc rewrite necessary in the near future.